### PR TITLE
aesc_htlc_SUITE (HTLC-based State Channel market logic)

### PR DESCRIPTION
**Pull request by uwiger**
See issue #3745 

This PR implements a State Channel HTLC contract and a test suite that runs through some scenarios.

The setup looks something like this:
```
                         +---------------------------------------------+                
                         |   Hub                                       |                
                         |                                             |                
                         |       +-+                         +-+       |                
                         |       |-+                         +|+       |                
                         +-------/----------------------------\--------+                
                                |                              |                        
                                /                              |                        
                               |                               \                        
                               /                                |                       
                              |                                 \                       
       +----------------------/--+                           +---|---------------------+
       | Alice              +|+  |                           | +-|                Bob  |
       |                    +-+  |                           | +-+                     |
       |                         |                           |                         |
       +-------------------------+                           +-------------------------+
```
When `Alice` wants to send tokens to `Bob`, she first sends a request via a generic message on the `Alice-Hub` channel. The request is forwarded by the `Hub` to `Bob` on the `Bob-Hub` channel. `Bob` generates a random secret, pairs it with the specifics of the transfer request, then hashes the result, producing a `hash_lock` value. This is passed back to `Alice` via generic message, forwarded by the `Hub`.

`Alice` then calls `new_send()` method of the HTLC contract. The `Hub` is automatically notified, as it is asked to co-sign. The `Hub` then performas a corresponding `new_receive()` call on the `Bob-Hub` channel. `Bob` retrieves the amount by passing the secret as a contract call argument to `receive()`. The `Hub` gets the secret in the co-signing request, and can then get its collateral back via a `collect()` call on the `Alice-Hub` channel.

A `timeout` parameter gives a window in which `Bob` can collect the amount. After `timeout + 3`, `Alice` (and `Hub`) can reclaim the amount (`Hub` will still get the specified fee from `Alice`). The timeout limits are staggered in order to avoid exploits via race conditions.

All contract calls are force-progressable on-chain, so for example, the `Hub` cannot deny a refund to `Alice` if the timeout has expired and the amount hasn't been collected. Likewise, `Alice` cannot keep `Hub` from collecting the amount.

### CODE REVIEW

**Comment by uwiger** [comment]
This is no longer WIP

**Comment by uwiger** [comment]
The PR includes av move of FATE calldata encode/decode from test utils to a production-side module.
This is part of an effort to support cleaner and more efficient contract calling from Erlang code.
At this point, the HTLC suite is content with getting FATE VM data types in return, since it doesn't rely on complex Sophia types.

File: test/contracts/channel_htlc.aes [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+  function is_active(x : status) : bool =
+    x == ACTIVE
+
+  function send_exists(id : hash) : bool =
+    Map.member(id, state.sends)
+
+  function recieve_exists(id : hash) : bool =
+    Map.member(id, state.receives)
+
+  function cc(ss : list(string)) : string =
```
Comment (hanssv) [inline test/contracts/channel_htlc.aes:218]: You can match `s :: ss` directly in the function clause

File: apps/aechannel/test/aesc_fsm_SUITE.erl [hunk 2]
```diff
@@ -5196,31 +5211,66 @@ create_contract(ContractName, InitArgs, Deposit,
     assert_empty_msgq(Debug),
     {Creator2, Acknowledger2, ContractPubkey}.
 
-call_contract(ContractId,
-              ContractName, FunName, FunArgs, Amount,
-              #{fsm := FsmC} = Caller, Acknowledger, Cfg) ->
-    Debug = get_debug(Cfg),
-    {ok, BinSrc} = aect_test_utils:read_contract(aect_test_utils:sophia_version(), ContractName),
+prepare_contract_create_args(ContractName, InitArgs, Deposit) ->
+    ct:log("SophiaVsn = ~p, IrisFate = ~p",
```
[the remark this answers is not in the document]
Comment (uwiger) [reply apps/aechannel/test/aesc_fsm_SUITE.erl:5215]: That was just so I could get my bearings, not being particularly used to decoding the sophia versions. :)

**Review by hanssv — APPROVED** [review]

**Review by seanhinde — APPROVED** [review]
Very cool

**Pull request merged**